In [1]:
#Pandas
import pandas as pd
import numpy as np 

*   pd.read_csv(path)  to load data set
*   In pandas the data type for such set is called as data frame
*   the first row is automatically treated as the column headers.

In [4]:
iris =   pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data")
print(type(iris))

<class 'pandas.core.frame.DataFrame'>


In [5]:
iris =   pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data")
iris

5.1  3.5  1.4  0.2     Iris-setosa
0    4.9  3.0  1.4  0.2     Iris-setosa
1    4.7  3.2  1.3  0.2     Iris-setosa
2    4.6  3.1  1.5  0.2     Iris-setosa
3    5.0  3.6  1.4  0.2     Iris-setosa
4    5.4  3.9  1.7  0.4     Iris-setosa
5    4.6  3.4  1.4  0.3     Iris-setosa
6    5.0  3.4  1.5  0.2     Iris-setosa
7    4.4  2.9  1.4  0.2     Iris-setosa
8    4.9  3.1  1.5  0.1     Iris-setosa
9    5.4  3.7  1.5  0.2     Iris-setosa
10   4.8  3.4  1.6  0.2     Iris-setosa
11   4.8  3.0  1.4  0.1     Iris-setosa
12   4.3  3.0  1.1  0.1     Iris-setosa
13   5.8  4.0  1.2  0.2     Iris-setosa
14   5.7  4.4  1.5  0.4     Iris-setosa
15   5.4  3.9  1.3  0.4     Iris-setosa
16   5.1  3.5  1.4  0.3     Iris-setosa
17   5.7  3.8  1.7  0.3     Iris-setosa
18   5.1  3.8  1.5  0.3     Iris-setosa
19   5.4  3.4  1.7  0.2     Iris-setosa
20   5.1  3.7  1.5  0.4     Iris-setosa
21   4.6  3.6  1.0  0.2     Iris-setosa
22   5.1  3.3  1.7  0.5     Iris-setosa
23   4.8  3.4  1.9  0.2     Iris-setosa
24   5.0  3.0  1.6  0.2     Iris-setosa
25   5.0  3.4  1.6  0.4     Iris-setosa
26   5.2  3.5  1.5  0.2     Iris-setosa
27   5.2  3.4  1.4  0.2     Iris-setosa
28   4.7  3.2  1.6  0.2     Iris-setosa
29   4.8  3.1  1.6  0.2     Iris-setosa
..   ...  ...  ...  ...             ...
119  6.9  3.2  5.7  2.3  Iris-virginica
120  5.6  2.8  4.9  2.0  Iris-virginica
121  7.7  2.8  6.7  2.0  Iris-virginica
122  6.3  2.7  4.9  1.8  Iris-virginica
123  6.7  3.3  5.7  2.1  Iris-virginica
124  7.2  3.2  6.0  1.8  Iris-virginica
125  6.2  2.8  4.8  1.8  Iris-virginica
126  6.1  3.0  4.9  1.8  Iris-virginica
127  6.4  2.8  5.6  2.1  Iris-virginica
128  7.2  3.0  5.8  1.6  Iris-virginica
129  7.4  2.8  6.1  1.9  Iris-virginica
130  7.9  3.8  6.4  2.0  Iris-virginica
131  6.4  2.8  5.6  2.2  Iris-virginica
132  6.3  2.8  5.1  1.5  Iris-virginica
133  6.1  2.6  5.6  1.4  Iris-virginica
134  7.7  3.0  6.1  2.3  Iris-virginica
135  6.3  3.4  5.6  2.4  Iris-virginica
136  6.4  3.1  5.5  1.8  Iris-virginica
137  6.0  3.0  4.8  1.8  Iris-virginica
138  6.9  3.1  5.4  2.1  Iris-virginica
139  6.7  3.1  5.6  2.4  Iris-virginica
140  6.9  3.1  5.1  2.3  Iris-virginica
141  5.8  2.7  5.1  1.9  Iris-virginica
142  6.8  3.2  5.9  2.3  Iris-virginica
143  6.7  3.3  5.7  2.5  Iris-virginica
144  6.7  3.0  5.2  2.3  Iris-virginica
145  6.3  2.5  5.0  1.9  Iris-virginica
146  6.5  3.0  5.2  2.0  Iris-virginica
147  6.2  3.4  5.4  2.3  Iris-virginica
148  5.9  3.0  5.1  1.8  Iris-virginica

[149 rows x 5 columns]

In [50]:
df = iris               #copies the reference so any change in one will reflect in the other 
df = iris.copy()        #copies the datas only

In [22]:
df.head()               #look at few initial enteries, by default it shows first five

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa

In [23]:
df.head(3)

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa

In [56]:
df.columns = ['sl', 'sw', 'pl', 'pw', 'flower_type']

In [25]:

print(df.shape)
print(df.dtypes)

(149, 5)
sl             float64
sw             float64
pl             float64
pw             float64
flower_type     object
dtype: object


In [26]:
df.describe()

sl          sw          pl          pw
count  149.000000  149.000000  149.000000  149.000000
mean     5.848322    3.051007    3.774497    1.205369
std      0.828594    0.433499    1.759651    0.761292
min      4.300000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.300000
50%      5.800000    3.000000    4.400000    1.300000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    2.500000

* count shows number of valid enteries. It wont be counted if there is a NaN in the entry.
* std is atandard deviation 
* 25%, 50% , 75% are the percentiles


In [28]:
df.sl
#df["sl"]   can be used also 

0      4.9
1      4.7
2      4.6
3      5.0
4      5.4
5      4.6
6      5.0
7      4.4
8      4.9
9      5.4
10     4.8
11     4.8
12     4.3
13     5.8
14     5.7
15     5.4
16     5.1
17     5.7
18     5.1
19     5.4
20     5.1
21     4.6
22     5.1
23     4.8
24     5.0
25     5.0
26     5.2
27     5.2
28     4.7
29     4.8
      ... 
119    6.9
120    5.6
121    7.7
122    6.3
123    6.7
124    7.2
125    6.2
126    6.1
127    6.4
128    7.2
129    7.4
130    7.9
131    6.4
132    6.3
133    6.1
134    7.7
135    6.3
136    6.4
137    6.0
138    6.9
139    6.7
140    6.9
141    5.8
142    6.8
143    6.7
144    6.7
145    6.3
146    6.5
147    6.2
148    5.9
Name: sl, Length: 149, dtype: float64

In [30]:
df.isnull()

sl     sw     pl     pw  flower_type
0    False  False  False  False        False
1    False  False  False  False        False
2    False  False  False  False        False
3    False  False  False  False        False
4    False  False  False  False        False
5    False  False  False  False        False
6    False  False  False  False        False
7    False  False  False  False        False
8    False  False  False  False        False
9    False  False  False  False        False
10   False  False  False  False        False
11   False  False  False  False        False
12   False  False  False  False        False
13   False  False  False  False        False
14   False  False  False  False        False
15   False  False  False  False        False
16   False  False  False  False        False
17   False  False  False  False        False
18   False  False  False  False        False
19   False  False  False  False        False
20   False  False  False  False        False
21   False  False  False  False        False
22   False  False  False  False        False
23   False  False  False  False        False
24   False  False  False  False        False
25   False  False  False  False        False
26   False  False  False  False        False
27   False  False  False  False        False
28   False  False  False  False        False
29   False  False  False  False        False
..     ...    ...    ...    ...          ...
119  False  False  False  False        False
120  False  False  False  False        False
121  False  False  False  False        False
122  False  False  False  False        False
123  False  False  False  False        False
124  False  False  False  False        False
125  False  False  False  False        False
126  False  False  False  False        False
127  False  False  False  False        False
128  False  False  False  False        False
129  False  False  False  False        False
130  False  False  False  False        False
131  False  False  False  False        False
132  False  False  False  False        False
133  False  False  False  False        False
134  False  False  False  False        False
135  False  False  False  False        False
136  False  False  False  False        False
137  False  False  False  False        False
138  False  False  False  False        False
139  False  False  False  False        False
140  False  False  False  False        False
141  False  False  False  False        False
142  False  False  False  False        False
143  False  False  False  False        False
144  False  False  False  False        False
145  False  False  False  False        False
146  False  False  False  False        False
147  False  False  False  False        False
148  False  False  False  False        False

[149 rows x 5 columns]

In [32]:
df.isnull().sum()

# number of null enteries in respective columns

sl             0
sw             0
pl             0
pw             0
flower_type    0
dtype: int64

In [71]:
#access a slice/part of the data

print(df.iloc[8:15, 2:4])       #works by position not index
df.loc[12]                      #works by label not position 

     pl   pw
12  1.1  0.1
13  1.2  0.2
14  1.5  0.4
15  1.3  0.4
16  1.4  0.3
17  1.7  0.3
18  1.5  0.3


sl                     4.3
sw                       3
pl                     1.1
pw                     0.1
flower_type    Iris-setosa
Name: 12, dtype: object

In [38]:
df.head()

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa

In [40]:
#droping by label

a = df.drop(1)  # copies the data deleting 1st row 

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa
5  4.6  3.4  1.4  0.3  Iris-setosa

In [51]:
df.drop(1, inplace = True)  # copies the data deleting 1st row in the dataset itself .  
df.head()

5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa
5  4.6  3.4  1.4  0.3  Iris-setosa

In [43]:
df.index

Int64Index([  0,   2,   3,   4,   5,   6,   7,   8,   9,  10,
            ...
            139, 140, 141, 142, 143, 144, 145, 146, 147, 148],
           dtype='int64', length=148)

In [45]:
df.index[1]  #shows the label/index of the position 1

2

In [52]:
#Droping by position 
df.drop(df.index[1], inplace = True)
df.head()



5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa
5  4.6  3.4  1.4  0.3  Iris-setosa
6  5.0  3.4  1.5  0.2  Iris-setosa

In [53]:
df.drop(df.index[[2,3]], inplace = True)         #remove more than one entries
df.head()

5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
6  5.0  3.4  1.5  0.2  Iris-setosa
7  4.4  2.9  1.4  0.2  Iris-setosa
8  4.9  3.1  1.5  0.1  Iris-setosa

In [58]:
df.sl > 5      #which rows has this condition true or false

0      False
3      False
6      False
7      False
8      False
9       True
10     False
11     False
12     False
13      True
14      True
15      True
16      True
17      True
18      True
19      True
20      True
21     False
22      True
23     False
24     False
25     False
26      True
27      True
28     False
29     False
30      True
31      True
32      True
33     False
       ...  
119     True
120     True
121     True
122     True
123     True
124     True
125     True
126     True
127     True
128     True
129     True
130     True
131     True
132     True
133     True
134     True
135     True
136     True
137     True
138     True
139     True
140     True
141     True
142     True
143     True
144     True
145     True
146     True
147     True
148     True
Name: sl, Length: 145, dtype: bool

In [61]:
#only those rows for which the condition is true 
df[df.sl > 5] 

sl   sw   pl   pw      flower_type
9    5.4  3.7  1.5  0.2      Iris-setosa
13   5.8  4.0  1.2  0.2      Iris-setosa
14   5.7  4.4  1.5  0.4      Iris-setosa
15   5.4  3.9  1.3  0.4      Iris-setosa
16   5.1  3.5  1.4  0.3      Iris-setosa
17   5.7  3.8  1.7  0.3      Iris-setosa
18   5.1  3.8  1.5  0.3      Iris-setosa
19   5.4  3.4  1.7  0.2      Iris-setosa
20   5.1  3.7  1.5  0.4      Iris-setosa
22   5.1  3.3  1.7  0.5      Iris-setosa
26   5.2  3.5  1.5  0.2      Iris-setosa
27   5.2  3.4  1.4  0.2      Iris-setosa
30   5.4  3.4  1.5  0.4      Iris-setosa
31   5.2  4.1  1.5  0.1      Iris-setosa
32   5.5  4.2  1.4  0.2      Iris-setosa
35   5.5  3.5  1.3  0.2      Iris-setosa
38   5.1  3.4  1.5  0.2      Iris-setosa
43   5.1  3.8  1.9  0.4      Iris-setosa
45   5.1  3.8  1.6  0.2      Iris-setosa
47   5.3  3.7  1.5  0.2      Iris-setosa
49   7.0  3.2  4.7  1.4  Iris-versicolor
50   6.4  3.2  4.5  1.5  Iris-versicolor
51   6.9  3.1  4.9  1.5  Iris-versicolor
52   5.5  2.3  4.0  1.3  Iris-versicolor
53   6.5  2.8  4.6  1.5  Iris-versicolor
54   5.7  2.8  4.5  1.3  Iris-versicolor
55   6.3  3.3  4.7  1.6  Iris-versicolor
57   6.6  2.9  4.6  1.3  Iris-versicolor
58   5.2  2.7  3.9  1.4  Iris-versicolor
60   5.9  3.0  4.2  1.5  Iris-versicolor
..   ...  ...  ...  ...              ...
119  6.9  3.2  5.7  2.3   Iris-virginica
120  5.6  2.8  4.9  2.0   Iris-virginica
121  7.7  2.8  6.7  2.0   Iris-virginica
122  6.3  2.7  4.9  1.8   Iris-virginica
123  6.7  3.3  5.7  2.1   Iris-virginica
124  7.2  3.2  6.0  1.8   Iris-virginica
125  6.2  2.8  4.8  1.8   Iris-virginica
126  6.1  3.0  4.9  1.8   Iris-virginica
127  6.4  2.8  5.6  2.1   Iris-virginica
128  7.2  3.0  5.8  1.6   Iris-virginica
129  7.4  2.8  6.1  1.9   Iris-virginica
130  7.9  3.8  6.4  2.0   Iris-virginica
131  6.4  2.8  5.6  2.2   Iris-virginica
132  6.3  2.8  5.1  1.5   Iris-virginica
133  6.1  2.6  5.6  1.4   Iris-virginica
134  7.7  3.0  6.1  2.3   Iris-virginica
135  6.3  3.4  5.6  2.4   Iris-virginica
136  6.4  3.1  5.5  1.8   Iris-virginica
137  6.0  3.0  4.8  1.8   Iris-virginica
138  6.9  3.1  5.4  2.1   Iris-virginica
139  6.7  3.1  5.6  2.4   Iris-virginica
140  6.9  3.1  5.1  2.3   Iris-virginica
141  5.8  2.7  5.1  1.9   Iris-virginica
142  6.8  3.2  5.9  2.3   Iris-virginica
143  6.7  3.3  5.7  2.5   Iris-virginica
144  6.7  3.0  5.2  2.3   Iris-virginica
145  6.3  2.5  5.0  1.9   Iris-virginica
146  6.5  3.0  5.2  2.0   Iris-virginica
147  6.2  3.4  5.4  2.3   Iris-virginica
148  5.9  3.0  5.1  1.8   Iris-virginica

[116 rows x 5 columns]

In [64]:
df[df.flower_type == "Iris-virginica"].describe()

sl         sw         pl        pw
count  50.00000  50.000000  50.000000  50.00000
mean    6.58800   2.974000   5.552000   2.02600
std     0.63588   0.322497   0.551895   0.27465
min     4.90000   2.200000   4.500000   1.40000
25%     6.22500   2.800000   5.100000   1.80000
50%     6.50000   3.000000   5.550000   2.00000
75%     6.90000   3.175000   5.875000   2.30000
max     7.90000   3.800000   6.900000   2.50000

In [73]:
df.head()


sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
6  5.0  3.4  1.5  0.2  Iris-setosa
7  4.4  2.9  1.4  0.2  Iris-setosa
8  4.9  3.1  1.5  0.1  Iris-setosa

In [76]:
print(df.iloc[1])      #non labelled by position 
print(df.loc[3])       #labelled  by index

sl                       5
sw                     3.6
pl                     1.4
pw                     0.2
flower_type    Iris-setosa
Name: 3, dtype: object
sl                       5
sw                     3.6
pl                     1.4
pw                     0.2
flower_type    Iris-setosa
Name: 3, dtype: object


In [80]:
#Add the row at the end with index 

df.loc[2] = [2.1,1.2,4.5,5.1, "iris"]
df.tail()

sl   sw   pl   pw     flower_type
145  6.3  2.5    5  1.9  Iris-virginica
146  6.5    3  5.2    2  Iris-virginica
147  6.2  3.4  5.4  2.3  Iris-virginica
148  5.9    3  5.1  1.8  Iris-virginica
2    2.1  1.2  4.5  5.1            iris

In [82]:
#resets the index 
df.reset_index()

index   sl   sw   pl   pw     flower_type
0        0  4.9    3  1.4  0.2     Iris-setosa
1        3    5  3.6  1.4  0.2     Iris-setosa
2        6    5  3.4  1.5  0.2     Iris-setosa
3        7  4.4  2.9  1.4  0.2     Iris-setosa
4        8  4.9  3.1  1.5  0.1     Iris-setosa
5        9  5.4  3.7  1.5  0.2     Iris-setosa
6       10  4.8  3.4  1.6  0.2     Iris-setosa
7       11  4.8    3  1.4  0.1     Iris-setosa
8       12  4.3    3  1.1  0.1     Iris-setosa
9       13  5.8    4  1.2  0.2     Iris-setosa
10      14  5.7  4.4  1.5  0.4     Iris-setosa
11      15  5.4  3.9  1.3  0.4     Iris-setosa
12      16  5.1  3.5  1.4  0.3     Iris-setosa
13      17  5.7  3.8  1.7  0.3     Iris-setosa
14      18  5.1  3.8  1.5  0.3     Iris-setosa
15      19  5.4  3.4  1.7  0.2     Iris-setosa
16      20  5.1  3.7  1.5  0.4     Iris-setosa
17      21  4.6  3.6    1  0.2     Iris-setosa
18      22  5.1  3.3  1.7  0.5     Iris-setosa
19      23  4.8  3.4  1.9  0.2     Iris-setosa
20      24    5    3  1.6  0.2     Iris-setosa
21      25    5  3.4  1.6  0.4     Iris-setosa
22      26  5.2  3.5  1.5  0.2     Iris-setosa
23      27  5.2  3.4  1.4  0.2     Iris-setosa
24      28  4.7  3.2  1.6  0.2     Iris-setosa
25      29  4.8  3.1  1.6  0.2     Iris-setosa
26      30  5.4  3.4  1.5  0.4     Iris-setosa
27      31  5.2  4.1  1.5  0.1     Iris-setosa
28      32  5.5  4.2  1.4  0.2     Iris-setosa
29      33  4.9  3.1  1.5  0.1     Iris-setosa
..     ...  ...  ...  ...  ...             ...
116    120  5.6  2.8  4.9    2  Iris-virginica
117    121  7.7  2.8  6.7    2  Iris-virginica
118    122  6.3  2.7  4.9  1.8  Iris-virginica
119    123  6.7  3.3  5.7  2.1  Iris-virginica
120    124  7.2  3.2    6  1.8  Iris-virginica
121    125  6.2  2.8  4.8  1.8  Iris-virginica
122    126  6.1    3  4.9  1.8  Iris-virginica
123    127  6.4  2.8  5.6  2.1  Iris-virginica
124    128  7.2    3  5.8  1.6  Iris-virginica
125    129  7.4  2.8  6.1  1.9  Iris-virginica
126    130  7.9  3.8  6.4    2  Iris-virginica
127    131  6.4  2.8  5.6  2.2  Iris-virginica
128    132  6.3  2.8  5.1  1.5  Iris-virginica
129    133  6.1  2.6  5.6  1.4  Iris-virginica
130    134  7.7    3  6.1  2.3  Iris-virginica
131    135  6.3  3.4  5.6  2.4  Iris-virginica
132    136  6.4  3.1  5.5  1.8  Iris-virginica
133    137    6    3  4.8  1.8  Iris-virginica
134    138  6.9  3.1  5.4  2.1  Iris-virginica
135    139  6.7  3.1  5.6  2.4  Iris-virginica
136    140  6.9  3.1  5.1  2.3  Iris-virginica
137    141  5.8  2.7  5.1  1.9  Iris-virginica
138    142  6.8  3.2  5.9  2.3  Iris-virginica
139    143  6.7  3.3  5.7  2.5  Iris-virginica
140    144  6.7    3  5.2  2.3  Iris-virginica
141    145  6.3  2.5    5  1.9  Iris-virginica
142    146  6.5    3  5.2    2  Iris-virginica
143    147  6.2  3.4  5.4  2.3  Iris-virginica
144    148  5.9    3  5.1  1.8  Iris-virginica
145      2  2.1  1.2  4.5  5.1            iris

[146 rows x 6 columns]

In [87]:
#reset the index droping previous indices
df.reset_index(drop = True, inplace = True)      #inorder to make changes in df itself set inplace to True 
df.head()

sl   sw   pl   pw  flower_type
0  4.9    3  1.4  0.2  Iris-setosa
1    5  3.6  1.4  0.2  Iris-setosa
2    5  3.4  1.5  0.2  Iris-setosa
3  4.4  2.9  1.4  0.2  Iris-setosa
4  4.9  3.1  1.5  0.1  Iris-setosa

In [88]:
df.index

RangeIndex(start=0, stop=146, step=1)

In [90]:
#delete a columnm 

df.drop("sl", axis = 1, inplace = True)      #axis 1 means look column wise


In [93]:
df.head()

sw   pl   pw  flower_type
0    3  1.4  0.2  Iris-setosa
1  3.6  1.4  0.2  Iris-setosa
2  3.4  1.5  0.2  Iris-setosa
3  2.9  1.4  0.2  Iris-setosa
4  3.1  1.5  0.1  Iris-setosa

In [97]:
del df["pl"]
df.describe()

pw      flower_type
count   146.0              146
unique   23.0                4
top       0.2  Iris-versicolor
freq     25.0               50

In [5]:
df = iris.copy()
df.columns = ['sl', 'sw', 'pl', 'pw', 'flower_type']
df.describe()


sl          sw          pl          pw
count  149.000000  149.000000  149.000000  149.000000
mean     5.848322    3.051007    3.774497    1.205369
std      0.828594    0.433499    1.759651    0.761292
min      4.300000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.300000
50%      5.800000    3.000000    4.400000    1.300000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    2.500000

In [100]:
df["diff_pl_pw"] = df["pl"] - df["pw"]
df.tail()

sl   sw   pl   pw     flower_type  diff_pl_pw
144  6.7  3.0  5.2  2.3  Iris-virginica         2.9
145  6.3  2.5  5.0  1.9  Iris-virginica         3.1
146  6.5  3.0  5.2  2.0  Iris-virginica         3.2
147  6.2  3.4  5.4  2.3  Iris-virginica         3.1
148  5.9  3.0  5.1  1.8  Iris-virginica         3.3

In [8]:
df.iloc[2:4,1:3] = np.nan
df.head()

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  NaN  NaN  0.2  Iris-setosa
3  5.0  NaN  NaN  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa

In [9]:
df.describe()

sl          sw          pl          pw
count  149.000000  147.000000  147.000000  149.000000
mean     5.848322    3.046939    3.806122    1.205369
std      0.828594    0.434048    1.750351    0.761292
min      4.300000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.300000
50%      5.800000    3.000000    4.400000    1.300000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    2.500000

sw and pl count is now 147 because NaN enteries are not counted.

In [12]:
df.dropna(inplace = True)      #drop rows having NaN enteris.

In [14]:
df.head()

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa
5  4.6  3.4  1.4  0.3  Iris-setosa
6  5.0  3.4  1.5  0.2  Iris-setosa

In [17]:
df.reset_index(drop = True, inplace = True)
df.head()

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  5.4  3.9  1.7  0.4  Iris-setosa
3  4.6  3.4  1.4  0.3  Iris-setosa
4  5.0  3.4  1.5  0.2  Iris-setosa

In [18]:
df.iloc[2:4,1:3] = np.nan
df.head()

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  5.4  NaN  NaN  0.4  Iris-setosa
3  4.6  NaN  NaN  0.3  Iris-setosa
4  5.0  3.4  1.5  0.2  Iris-setosa

In [20]:
#fill NaN enteries

df.sw.fillna(df.sw.mean(), inplace = True )
df.pl.fillna(df.pl.mean(), inplace = True )
df.head()

sl        sw        pl   pw  flower_type
0  4.9  3.000000  1.400000  0.2  Iris-setosa
1  4.7  3.200000  1.300000  0.2  Iris-setosa
2  5.4  3.038621  3.837241  0.4  Iris-setosa
3  4.6  3.038621  3.837241  0.3  Iris-setosa
4  5.0  3.400000  1.500000  0.2  Iris-setosa

In [31]:
#String based Data
df["Gender"] = "Female"
df.iloc[2:10, 5] = "Male"
df.head()

sl        sw        pl   pw  flower_type  Gender
0  4.9  3.000000  1.400000  0.2  Iris-setosa  Female
1  4.7  3.200000  1.300000  0.2  Iris-setosa  Female
2  5.4  3.038621  3.837241  0.4  Iris-setosa    Male
3  4.6  3.038621  3.837241  0.3  Iris-setosa    Male
4  5.0  3.400000  1.500000  0.2  Iris-setosa    Male

In [32]:
def f(s):
    if s == "Male":
        return 0
    else:
        return 1
df["Sex"] =df.Gender.apply(f)       #apply rule/function f
df.head()

sl        sw        pl   pw  flower_type  Gender  Sex
0  4.9  3.000000  1.400000  0.2  Iris-setosa  Female    1
1  4.7  3.200000  1.300000  0.2  Iris-setosa  Female    1
2  5.4  3.038621  3.837241  0.4  Iris-setosa    Male    0
3  4.6  3.038621  3.837241  0.3  Iris-setosa    Male    0
4  5.0  3.400000  1.500000  0.2  Iris-setosa    Male    0

In [1]:
df.dtypes

NameError: name 'df' is not defined